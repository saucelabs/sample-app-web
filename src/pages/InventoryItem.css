.inventory_item_container {
  padding: 40px 15px;
}

.inventory_details_container {
  display: flex;
  max-width: 1280px;
  margin: 0 auto;
  border: 2px solid #ededef;
}

.inventory_details_container .inventory_details_img_container {
  flex: 2;
  display: block;
}

.inventory_details_container .inventory_details_img {
  flex: 1;
  height: 100%;
  width: 100%;
}

.inventory_details_container .inventory_details_desc_container {
  padding: 20px 20px 0;
  flex: 3;
}

.inventory_details_container .inventory_details_name {
  font-family: "Roboto", Arial, Helvetica, sans-serif;
  font-size: 20px;
  font-weight: 500;
  color: #e2231a;
  margin-bottom: 15px;
}

.inventory_details_container .inventory_details_desc {
  font-family: "Roboto", Arial, Helvetica, sans-serif;
  font-size: 14px;
  color: #777777;
  margin-bottom: 30px;
}

.inventory_details_container .inventory_details_price {
  border-top: 2px solid #ededef;
  color: #e2231a;
  font-family: "Roboto", Arial, Helvetica, sans-serif;
  font-size: 22px;
  display: inline-block;
  padding-top: 10px;
}

.inventory_details_container .inventory_details_desc_container .btn {
  margin-top: 30px;
  display: block;
}

.inventory_details_container .large_size {
  width: 50%;
}

@media only screen and (max-width: 900px) {
  .inventory_details_container .large_size {
    width: 100%;
  }

  .inventory_details_container .inventory_details_img_container {
    flex: 1;
  }

  .inventory_details_container .inventory_details_desc_container {
    flex: 2;
  }

  .inventory_details_container .inventory_details_desc {
    margin-bottom: 15px;
  }

  .inventory_details_container .inventory_details_desc_container .btn {
    display: inline;
    margin-left: 15px;
    margin-top: 15px;
    width: 160px;
  }
}

@media only screen and (max-width: 480px) {
  .inventory_details_container {
    border: none;
    flex-direction: column;
  }

  .inventory_details .left_component {
    width: 100%;
  }

  .inventory_details_container .inventory_item_container {
    padding: 40px 35px;
  }

  .inventory_details_container .inventory_details_desc_container {
    flex: 1;
    padding: 20px 0;
  }

  .inventory_details_container .inventory_details_desc {
    margin-bottom: 20px;
  }

  .inventory_details_container .inventory_details_desc_container .btn {
    display: block;
    margin-left: 0;
    margin-top: 30px;
    width: 100%;
  }
}
