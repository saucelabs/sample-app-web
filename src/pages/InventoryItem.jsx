import React, { useEffect, useState } from "react";
import { withRouter } from "react-router-dom";
import { isProblemUser } from "../utils/Credentials";
import { ROUTES } from "../utils/Constants";
import { ShoppingCart } from "../utils/shopping-cart";
import { InventoryData } from "../utils/InventoryData";
import HeaderContainer from "../components/HeaderContainer";
import Button, { BUTTON_SIZES, BUTTON_TYPES } from "../components/Button";
import SwagLabsFooter from "../components/Footer";
import "./InventoryItem.css";

const InventoryItem = (props) => {
  useEffect(() => {
    window.scrollTo(0, 0);
  }, []);
  const { history } = props;
  // Get our queryparams now
  const queryParams = new URLSearchParams(window.location.search);
  let inventoryId = -1;
  let item;

  /* istanbul ignore else */
  if (queryParams.has("id")) {
    inventoryId = parseInt(queryParams.get("id"));
  }

  if (inventoryId >= 0 && InventoryData.length > inventoryId) {
    item = InventoryData[inventoryId];
  } else {
    item = {
      name: "ITEM NOT FOUND",
      desc: `We're sorry, but your call could not be completed as dialled.
          Please check your number, and try your call again.
          If you are in need of assistance, please dial 0 to be connected with an operator.
          This is a recording.
          4 T 1.`,
      image_url: "sl-404.jpg",
      price: "âˆš-1",
    };
  }

  item.id = inventoryId;

  const [itemInCart, setItemInCart] = useState(
    ShoppingCart.isItemInCart(inventoryId)
  );
  /**
   * @TODO:
   * This can't be tested yet because enzyme currently doesn't support ReactJS17,
   * see https://github.com/enzymejs/enzyme/issues/2429.
   * This means we can't fully mount the component and test all rendered components
   * and functions
   */
  /* istanbul ignore next */
  const goBack = () => {
    history.push(ROUTES.INVENTORY);
  };
  /**
   * @TODO:
   * This can't be tested yet because enzyme currently doesn't support ReactJS17,
   * see https://github.com/enzymejs/enzyme/issues/2429.
   * This means we can't fully mount the component and test all rendered components
   * and functions
   */
  /* istanbul ignore next */
  const addToCart = (itemId) => {
    if (isProblemUser()) {
      // Bail out now, don't add to cart if the item ID is odd
      if (itemId % 2 === 1) {
        return;
      }
    }

    ShoppingCart.addItem(itemId);
    setItemInCart(true);
  };
  /**
   * @TODO:
   * This can't be tested yet because enzyme currently doesn't support ReactJS17,
   * see https://github.com/enzymejs/enzyme/issues/2429.
   * This means we can't fully mount the component and test all rendered components
   * and functions
   */
  /* istanbul ignore next */
  const removeFromCart = (itemId) => {
    if (isProblemUser()) {
      // Bail out now, don't remove from cart if the item ID is even
      if (itemId % 2 === 0) {
        return;
      }
    }

    ShoppingCart.removeItem(itemId);
    setItemInCart(false);
  };
  /**
   * @TODO:
   * This can't be tested yet because enzyme currently doesn't support ReactJS17,
   * see https://github.com/enzymejs/enzyme/issues/2429.
   * This means we can't fully mount the component and test all rendered components
   * and functions
   */
  /* istanbul ignore next */
  const ButtonType = ({ id, item, itemInCart }) => {
    const label = itemInCart ? "Remove" : "Add to cart";
    const onClick = itemInCart ? () => removeFromCart(id) : () => addToCart(id);
    const type = itemInCart ? BUTTON_TYPES.SECONDARY : BUTTON_TYPES.PRIMARY;
    const testId = `${label}-${item}`.replace(/\s+/g, "-").toLowerCase();

    return (
      <Button
        customClass="btn_inventory"
        label={label}
        onClick={onClick}
        size={BUTTON_SIZES.SMALL}
        testId={testId}
        type={type}
      />
    );
  };

  return (
    <div id="page_wrapper">
      <div id="contents_wrapper">
        <HeaderContainer
          customClass="inventory_details"
          secondaryLeftComponent={
            <Button
              customClass="inventory_details_back_button"
              label="Back to products"
              onClick={goBack}
              type={BUTTON_TYPES.BACK}
              testId="back-to-products"
            />
          }
        />
        <div id="inventory_item_container" className="inventory_item_container">
          <div className="inventory_details">
            <div className="inventory_details_container">
              <div className="inventory_details_img_container">
                <img
                  alt={item.name}
                  className="inventory_details_img"
                  src={require(`../assets/img/${item.image_url}`).default}
                />
              </div>
              <div className="inventory_details_desc_container">
                <div className="inventory_details_name large_size">
                  {item.name}
                </div>
                <div className="inventory_details_desc large_size">
                  {item.desc}
                </div>
                <div className="inventory_details_price">${item.price}</div>
                <ButtonType
                  id={item.id}
                  itemInCart={itemInCart}
                  item={item.name}
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <SwagLabsFooter />
    </div>
  );
};

export default withRouter(InventoryItem);
