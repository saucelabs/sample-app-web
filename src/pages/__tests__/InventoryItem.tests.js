import React, { useState as useStateMock } from "react";
import { shallow } from "enzyme";
import InventoryItem from "../InventoryItem";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
  useEffect: (f) => f(),
}));

let props;
let location;

describe("InventoryItem", () => {
  const setState = jest.fn();
  const spyScrollTo = jest.fn();

  beforeEach(() => {
    props = {
      history: {
        push: jest.fn(),
      },
    };
    location = {
      ...window.location,
      search: "?id=4",
    };
    Object.defineProperty(window, "location", {
      writable: true,
      value: location,
    });
    Object.defineProperty(global.window, "scrollTo", { value: spyScrollTo });
    Object.defineProperty(global.window, "scrollY", { value: 1 });
    spyScrollTo.mockClear();
    useStateMock.mockImplementation((init) => [init, setState]);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render with default props", () => {
    const component = shallow(<InventoryItem.WrappedComponent {...props} />);

    expect(component).toMatchSnapshot();
    expect(spyScrollTo).toHaveBeenCalledTimes(1);
    expect(setState).toHaveBeenCalledTimes(0);
  });

  it("should render with error data if a not known product is opened", () => {
    location = {
      ...window.location,
      search: "?id=999",
    };
    Object.defineProperty(window, "location", {
      writable: true,
      value: location,
    });
    const component = shallow(<InventoryItem.WrappedComponent {...props} />);

    expect(component).toMatchSnapshot();
  });
  //
  // it("should be able to open the details page when the swag image is clicked", () => {
  //   const component = shallow(
  //     <InventoryListItem.WrappedComponent {...props} />
  //   );
  //
  //   const imageLink = component.find(`#item_${props.id}_img_link`).at(0);
  //   imageLink.simulate("click", {
  //     preventDefault() {},
  //   });
  //   expect(props.history.push).toBeCalledWith(
  //     `/inventory-item.html?id=${props.id}`
  //   );
  // });
  //
  // it("should be able to open the details page when the swag item title is clicked", () => {
  //   const component = shallow(
  //     <InventoryListItem.WrappedComponent {...props} />
  //   );
  //
  //   const imageLink = component.find(`#item_${props.id}_title_link`).at(0);
  //   imageLink.simulate("click", {
  //     preventDefault() {},
  //   });
  //   expect(props.history.push).toBeCalledWith(
  //     `/inventory-item.html?id=${props.id}`
  //   );
  // });
  //
  // it("should be able to set the image url and id for a problem user", () => {
  //   const isProblemUserSpy = jest.spyOn(Credentials, "isProblemUser");
  //   isProblemUserSpy.mockReturnValue(true);
  //   const component = shallow(
  //     <InventoryListItem.WrappedComponent {...props} />
  //   );
  //   const imageLink = component.find(`#item_${props.id}_title_link`).at(0);
  //   imageLink.simulate("click", {
  //     preventDefault() {},
  //   });
  //
  //   expect(props.history.push).toBeCalledWith(
  //     `/inventory-item.html?id=${props.id + 1}`
  //   );
  //   expect(setState).toHaveBeenCalledWith(
  //     `${props.image_url}WithGarbageOnItToBreakTheUrl`
  //   );
  //
  //   isProblemUserSpy.mockClear();
  // });
});
