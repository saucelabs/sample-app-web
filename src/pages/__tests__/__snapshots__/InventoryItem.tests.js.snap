// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryItem should render with default props 1`] = `
<div
  id="page_wrapper"
>
  <div
    id="contents_wrapper"
  >
    <HeaderContainer
      customClass="inventory_details"
      secondaryLeftComponent={
        <Button
          customClass="inventory_details_back_button"
          label="Back to products"
          onClick={[Function]}
          size="large"
          testId="back-to-products"
          type="secondary back"
        />
      }
    />
    <div
      className="inventory_item_container"
      id="inventory_item_container"
    >
      <div
        className="inventory_details"
      >
        <div
          className="inventory_details_container"
        >
          <div
            className="inventory_details_img_container"
          >
            <img
              alt="Sauce Labs Backpack"
              className="inventory_details_img"
            />
          </div>
          <div
            className="inventory_details_desc_container"
          >
            <div
              className="inventory_details_name large_size"
            >
              Sauce Labs Backpack
            </div>
            <div
              className="inventory_details_desc large_size"
            >
              carry.allTheThings() with the sleek, streamlined Sly Pack that melds uncompromising style with unequaled laptop and tablet protection.
            </div>
            <div
              className="inventory_details_price"
            >
              $
              29.99
            </div>
            <ButtonType
              id={4}
              item="Sauce Labs Backpack"
              itemInCart={false}
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <SwagLabsFooter />
</div>
`;

exports[`InventoryItem should render with error data if a not known product is opened 1`] = `
<div
  id="page_wrapper"
>
  <div
    id="contents_wrapper"
  >
    <HeaderContainer
      customClass="inventory_details"
      secondaryLeftComponent={
        <Button
          customClass="inventory_details_back_button"
          label="Back to products"
          onClick={[Function]}
          size="large"
          testId="back-to-products"
          type="secondary back"
        />
      }
    />
    <div
      className="inventory_item_container"
      id="inventory_item_container"
    >
      <div
        className="inventory_details"
      >
        <div
          className="inventory_details_container"
        >
          <div
            className="inventory_details_img_container"
          >
            <img
              alt="ITEM NOT FOUND"
              className="inventory_details_img"
            />
          </div>
          <div
            className="inventory_details_desc_container"
          >
            <div
              className="inventory_details_name large_size"
            >
              ITEM NOT FOUND
            </div>
            <div
              className="inventory_details_desc large_size"
            >
              We're sorry, but your call could not be completed as dialled.
          Please check your number, and try your call again.
          If you are in need of assistance, please dial 0 to be connected with an operator.
          This is a recording.
          4 T 1.
            </div>
            <div
              className="inventory_details_price"
            >
              $
              âˆš-1
            </div>
            <ButtonType
              id={999}
              item="ITEM NOT FOUND"
              itemInCart={false}
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <SwagLabsFooter />
</div>
`;
